File: src/ros2_cmake/src/service_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "ros2_interface/srv/ex_custom_srv.hpp"
#include <chrono>
#include <cstdlib>
#include <memory>

using namespace std::chrono_literals;

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);

    // 실행 시 인자 체크
    if (argc != 3) {
        RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "usage: add_two_ints_client X Y");
        return 1;
    }

    // 노드 생성 (이름: add_two_ints_client)
    std::shared_ptr<rclcpp::Node> node = 
        rclcpp::Node::make_shared("add_two_ints_client");

    // 서비스 클라이언트 생성
    // "add_two_ints" 라는 이름의 서비스에 요청을 보냄
    rclcpp::Client<ros2_interface::srv::ExCustomSrv>::SharedPtr client =
        node->create_client<ros2_interface::srv::ExCustomSrv>("add_two_ints");

    // 요청 객체 생성
    auto request = std::make_shared<ros2_interface::srv::ExCustomSrv::Request>();

    // 입력값을 숫자로 변환해서 request에 넣기
    request->a = atoll(argv[1]);
    request->b = atoll(argv[2]);

    // 서비스 서버가 준비될 때까지 기다림
    while (!client->wait_for_service(1s)) {
        // ROS가 종료 상태라면
        if (!rclcpp::ok()) {
            RCLCPP_ERROR(rclcpp::get_logger("rclcpp"),
                         "Interrupted while waiting for the service. Exiting.");
            return 0;
        }

        // 아직 서비스가 없으면 계속 기다림
        RCLCPP_INFO(rclcpp::get_logger("rclcpp"),
                    "service not available, waiting again...");
    }

    // 비동기 요청 보내기
    auto result = client->async_send_request(request);

    // 결과 올 때까지 대기
    if (rclcpp::spin_until_future_complete(node, result) == rclcpp::FutureReturnCode::SUCCESS)
    {
        // 결과 출력 (sum 값)
        RCLCPP_INFO(rclcpp::get_logger("rclcpp"),
                    "Sum: %ld", result.get()->sum);
    } 
    else {
        // 실패했을 때
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"),
                     "Failed to call service add_two_ints");
    }

    // ROS2 종료
    rclcpp::shutdown();
    return 0;
}

File: src/ros2_python/ros2_python/service_client.py
import sys
from ros2_interface.srv import ExCustomSrv
import rclpy
from rclpy.node import Node

# 서비스 클라이언트 노드 정의
class CustomClientAsync(Node):
    def __init__(self):
        # 노드 이름 설정
        super().__init__('minimal_client_async')

        # 서비스 클라이언트 생성
        # 서비스 이름: add_two_ints (서버와 동일해야 함)
        self.cli = self.create_client(ExCustomSrv, 'add_two_ints')

        # 서비스 서버가 준비될 때까지 기다림
        while not self.cli.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('service not available, waiting again...')

        # 요청 객체 생성
        self.req = ExCustomSrv.Request()

    # 서비스 요청 보내는 함수
    def send_request(self, a, b):
        self.req.a = a
        self.req.b = b

        # 비동기 요청 전송 (future 반환)
        return self.cli.call_async(self.req)


def main():
   rclpy.init()

   # 클라이언트 노드 생성
   custom_client = CustomClientAsync()

   # 터미널 입력값 가져오기
   future = custom_client.send_request(int(sys.argv[1]), int(sys.argv[2]))

   # 응답 올 때까지 기다림
   rclpy.spin_until_future_complete(custom_client, future)

   # 결과 받아오기
   response = future.result()

   # 결과 출력
   custom_client.get_logger().info(
       'Result of add_two_ints: for %d + %d = %d' % (
           int(sys.argv[1]), int(sys.argv[2]), response.sum))

   # 노드 종료
   custom_client.destroy_node()

   # ROS2 종료
   rclpy.shutdown()
   

if __name__ == '__main__':
   main()